res/res_stir_shaken/verification.c

```diff
@@ -0,0 +1,1033 @@
+		expires.tv_sec, current_time.tv_sec, expires.tv_sec - current_time.tv_sec);
+
+	res = (ast_tvcmp(current_time, expires) == -1 ? 0 : 1);
+	SCOPE_EXIT_RTN_VALUE(res , "entry was %sexpired\n", res ? "" : "not ");
+}
+
+static enum ast_stir_shaken_vs_response_code check_cert(
+	struct ast_stir_shaken_vs_ctx * ctx)
+{
+	ASN1_OCTET_STRING *tn_exten;
+	int res = 0;
+	SCOPE_ENTER(3, "%s: Validating cert '%s'\n", ctx->tag, ctx->public_url);
+
+	ast_rwlock_rdlock(&trusted_cert_store.store_lock);
+	ast_trace(3,"%s: Checking ctx against CA ctx\n", ctx->tag);
+	res = ast_crypto_is_cert_trusted(trusted_cert_store.store, ctx->xcert);
+	ast_rwlock_unlock(&trusted_cert_store.store_lock);
+	if (!res) {
+		SCOPE_EXIT_LOG_RTN_VALUE(AST_STIR_SHAKEN_VS_CERT_NOT_TRUSTED,
+			LOG_ERROR, "%s: Cert '%s' not trusted\n",
+			ctx->tag, ctx->public_url);
+	}
+
+	ast_trace(3,"%s: Attempting to get the raw pubkey\n", ctx->tag);
+	ctx->raw_key_len = ast_crypto_get_raw_pubkey_from_cert(ctx->xcert,
+		&ctx->raw_key);
+	if (ctx->raw_key_len <= 0) {
+		SCOPE_EXIT_LOG_RTN_VALUE(AST_STIR_SHAKEN_VS_NO_RAW_KEY,
+			LOG_ERROR, "%s: Unable to extract raw public key from '%s'\n",
+			ctx->tag, ctx->public_url);
+	}
+
+	ast_trace(3,"%s: Checking cert '%s' validity dates\n",
+		ctx->tag, ctx->public_url);
+	if (!ast_crypto_is_cert_time_valid(ctx->xcert, ctx->validity_check_time)) {
+		SCOPE_EXIT_LOG_RTN_VALUE(AST_STIR_SHAKEN_VS_CERT_DATE_INVALID,
+			LOG_ERROR, "%s: Cert '%s' dates not valid\n",
+			ctx->tag, ctx->public_url);
+	}
+
+	ast_trace(3,"%s: Checking cert '%s' TNAuth extensions\n",
+		ctx->tag, ctx->public_url);
+	tn_exten = ast_crypto_get_cert_extension_data(ctx->xcert, ss_get_tn_auth_nid(), NULL);
+	if (!tn_exten) {
+		SCOPE_EXIT_LOG_RTN_VALUE(AST_STIR_SHAKEN_VS_CERT_NO_TN_AUTH_EXT,
+			LOG_ERROR, "%s: Cert '%s' doesn't have a TNAuthList extension\n",
+			ctx->tag, ctx->public_url);
+		return 0;
+	}
+
+	SCOPE_EXIT_RTN_VALUE(AST_STIR_SHAKEN_VS_SUCCESS, "%s: Cert '%s' is valid\n",
+		ctx->tag, ctx->public_url);
+}
+
+
+static enum ast_stir_shaken_vs_response_code retrieve_cert_from_url(
+		struct ast_stir_shaken_vs_ctx *ctx)
+{
+	FILE *cert_file;
+	long http_code;
+	int rc = 0;
+	enum ast_stir_shaken_vs_response_code vs_rc;
+	RAII_VAR(struct curl_header_data *, header_data,
+		ast_calloc(1, sizeof(*header_data)), curl_header_data_free);
+	RAII_VAR(struct curl_write_data *, write_data,
+		ast_calloc(1, sizeof(*write_data)), curl_write_data_free);
+	RAII_VAR(struct curl_open_socket_data *, open_socket_data,
+		ast_calloc(1, sizeof(*open_socket_data)), curl_open_socket_data_free);
+
+	const char *cache_control;
+	const char *expires;
+	SCOPE_ENTER(2, "%s: Attempting to retrieve '%s' from net\n",
+		ctx->tag, ctx->public_url);
+
+	header_data->debug_info = ast_strdup(ctx->public_url);
+	write_data->debug_info = ast_strdup(ctx->public_url);
+	write_data->max_download_bytes = 8192;
+	write_data->stream_buffer = NULL;
+	open_socket_data->debug_info = ast_strdup(ctx->public_url);
```

Comment:
Lots of places here when we aren't checking for allocation failures

Reply:
done